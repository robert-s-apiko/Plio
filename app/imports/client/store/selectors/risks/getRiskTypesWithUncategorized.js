import { createSelector } from 'reselect';
import { lenses, lensEqById, withUncategorized } from 'plio-util';

import { getRiskTypes } from '../riskTypes';
import { createUncategorizedType } from '../../../../ui/react/risks/helpers';

const getRisks = (_, { risks }) => risks;

// selector(risks: Array, types: Array) => Array
const selector = (risks, types) => withUncategorized(
  // (lens: Lens) => (type: Object) => (risk: Object) => Boolean
  lensEqById(lenses.typeId),
  lenses.typeId,
  lenses.risks,
  createUncategorizedType({ risks, types }),
  risks,
  types,
);

export default createSelector([
  getRisks,
  getRiskTypes,
], selector);
