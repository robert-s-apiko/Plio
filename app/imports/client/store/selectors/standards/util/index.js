export { default as withUncategorized } from './withUncategorized';
