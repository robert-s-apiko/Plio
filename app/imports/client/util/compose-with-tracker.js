import { Tracker } from 'meteor/tracker';
import { compose } from 'ramda';

function getTrackerLoader(loaderFunc) {
  return (props, onData, env) => {
    let trackerCleanup = () => null;

    const handler = Tracker.nonreactive(() =>
      Tracker.autorun(() => {
        // Store clean-up function if provided.
        trackerCleanup = loaderFunc(props, onData, env) || (() => null);
      }));

    return () => {
      trackerCleanup();
      return handler.stop();
    };
  };
}

function composeWithTracker(loadFunc) {
  return function (component) {
    return compose(getTrackerLoader(loadFunc))(component);
  };
}

export default composeWithTracker;
