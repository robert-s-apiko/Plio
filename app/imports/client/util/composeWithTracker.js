import { Tracker } from 'meteor/tracker';
import { compose } from '@storybook/react-komposer';

function getTrackerLoader(loaderFunc) {
  return (props, onData, env) => {
    let trackerCleanup = () => null;

    const handler = Tracker.nonreactive(() =>
      Tracker.autorun(() => {
        // Store clean-up function if provided.
        trackerCleanup = loaderFunc(props, onData, env) || (() => null);
      }));

    return () => {
      trackerCleanup();
      return handler.stop();
    };
  };
}

function composeWithTracker(loadFunc, options) {
  return function (component) {
    // explicitly pass { withRef: false }
    // otherwise react will throw error
    // that the component tries to access
    // a ref of the stateless component
    const opts = Object.assign({ withRef: false }, options);

    return compose(getTrackerLoader(loadFunc), opts)(component);
  };
}

export default composeWithTracker;
