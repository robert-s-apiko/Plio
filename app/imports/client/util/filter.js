import { ifElse, equals, compose, filter, nthArg, complement } from 'ramda';

import { isDeleted } from './object-relation';

// (filterDeleted: Boolean, items: Array) => Array
export const filterByDeleted = ifElse(
  equals(true),
  compose(filter(isDeleted), nthArg(1)),
  compose(filter(complement(isDeleted)), nthArg(1)),
);
