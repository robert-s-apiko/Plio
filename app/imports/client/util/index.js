export { default as composeWithTracker } from './composeWithTracker';
