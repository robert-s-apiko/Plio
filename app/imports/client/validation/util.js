import { FORM_ERROR } from 'final-form';
import { forEachObjIndexed, curry, isNil } from 'ramda';

export const required = curry((label, value) => !isNil(value) ? undefined : `${label} required`);

export const validate = rules => (values) => {
  const errors = new Set();

  forEachObjIndexed((value, key) => {
    if (rules[key]) {
      const error = rules[key](value, values);
      if (error) errors.add(error);
    }
  }, values);

  if (!errors.size) return undefined;

  return { [FORM_ERROR]: [...errors].join('\n') };
};
