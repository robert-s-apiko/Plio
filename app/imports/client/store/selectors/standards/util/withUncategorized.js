import { set, filter, propEq, append, compose, map } from 'ramda';
import { lenses, getStandardsLength } from '../../../../util';

// (prop: String, standards: Array) => (item: Object) => Object
export const setFilteredStandards = (prop, standards) => item => set(
  lenses.standards,
  filter(propEq(prop, item._id), standards),
  item,
);

// (prop: String, uncategorized: Object, standards: Array, items: Array) => Array
export default (prop, uncategorized, standards, items) => compose(
  filter(getStandardsLength),
  append(uncategorized),
  map(setFilteredStandards(prop, standards)),
)(items);
