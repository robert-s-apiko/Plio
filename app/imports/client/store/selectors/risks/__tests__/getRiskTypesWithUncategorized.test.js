import { drop } from 'ramda';
import getRiskTypesWithUncategorized from '../getRiskTypesWithUncategorized';
import { getInitialState } from '../../../util/tests';

describe('getRiskTypesWithUncategorized', () => {
  it('matches snapshot', () => {
    const state = getInitialState({});
    const props = {
      risks: state.collections.risks,
    };

    expect(getRiskTypesWithUncategorized(state, props)).toMatchSnapshot();
  });

  it('should correctly memoize', () => {
    const state = getInitialState({});
    let props = {
      risks: state.collections.risks,
    };
    getRiskTypesWithUncategorized(state, props);
    getRiskTypesWithUncategorized(state, props);
    expect(getRiskTypesWithUncategorized.recomputations()).toBe(1);

    props = {
      risks: drop(1, props.risks),
    };

    getRiskTypesWithUncategorized(state, props);
    expect(getRiskTypesWithUncategorized.recomputations()).toBe(2);
  });
});
