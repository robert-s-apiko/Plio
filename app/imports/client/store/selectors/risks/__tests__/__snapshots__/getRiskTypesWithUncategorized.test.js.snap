// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getRiskTypesWithUncategorized matches snapshot 1`] = `
Array [
  Object {
    "_id": 15,
    "risks": Array [
      Object {
        "_id": 10,
        "departmentsIds": Array [
          16,
        ],
        "standardsIds": Array [
          1,
        ],
        "typeId": 15,
      },
      Object {
        "_id": 11,
        "departmentsIds": Array [
          16,
          17,
        ],
        "standardsIds": Array [
          1,
          2,
        ],
        "typeId": 15,
      },
      Object {
        "_id": 12,
        "deletedAt": 1513074233019,
        "departmentsIds": Array [
          17,
        ],
        "isDeleted": true,
        "standardsIds": Array [
          5,
        ],
        "typeId": 15,
      },
    ],
  },
  Object {
    "_id": "RiskTypes.Uncategorized",
    "organizationId": undefined,
    "risks": Array [
      Object {
        "_id": 13,
        "deletedAt": 1513074233018,
        "isDeleted": true,
        "standardsIds": Array [
          1,
        ],
      },
      Object {
        "_id": 14,
        "standardsIds": Array [
          2,
        ],
      },
    ],
    "title": "Uncategorized",
  },
]
`;
